### PR TITLE
Rust dev

### PR DESCRIPTION
**Pull request by fredclausen**
Rewrite of acars_router from python in to Rust.

~~**PR is opened to begin testing CI, not merge ready!!!**~~
**PR is ready for testing**

## Major functionality:

- [x] Listen UDP
- [x] Send UDP
- [x] Listen TCP
- [x] Send TCP
- [x] Listen ZMQ
- [x] Send ZMQ
- [x] Docker build
- [x] GitHub CI verification
- [x] Pass unit tests
- [x] Log statistics
- [x] Over ride station name
- [x] Drop in replacement, ie all command line/env flags are respected and operate the same

## Breaking Changes

* Move docker base image from python to SDR-E Base. Will require a change to the SDR-E Base Image CI to rebuild the container (sdr-enthusiasts/docker-baseimage/pull/92)
* tmpfs build step needed for common GitHub actions (sdr-enthusiasts/common-github-workflows/pull/9)
* ~~`--add-proxy-id` is now `--dont-add-proxy-id`~~ **REVERTED TO PREVIOUS BEHAVIOR**
* ~~`--verbose` supports the legacy numbering scheme to indicate logging level, but documentation is updated to indicate~~  ~~`info`/`debug`/`trace` as the preferred convention.~~ `-v/-vv` for `debug/trace`. `AR_VERBOSITY` respects old legacy and `info/debug/trace` convention.
* `--receive-zmq-acars` option added. 
* `--reassembly-window` option added.

### CODE REVIEW

**Comment by fredclausen** [comment]
@mikenye at this point I think it's time to start letting people test the rewrite with your blessing. The best I can tell it's as close to a 1:1 rewrite of the functionality of the python version, and to weed out the bugs we'll have to let people use it. It should require zero configuration change. All command line and/or env options are respected.

If you want to build and test before we let users play with it you can check out the `rust-dev` branch and build the docker file. It'll do all the work, no tool chain on your computer required. If you want to build outside of docker make sure Rust is installed, cd in to the repo's root dir and do:

```shell
cargo build
```

or

```shell
cargo run -- <command line flags for acars_router>
```

First one just builds it, the second builds and then runs the resulting binary.

I suggest, for testing purposes because this is a brand new program, that we temporarily add in to the on pr action `push: true` and push it to a `test` tag or something like that. This would enable us to have users start playing with it. We can remove the push before the final merge and also delete the `test` image after we're all done. Also, I've removed the python build Dockerfile but I think it might be worth adding that in as a seperate dockerfile and building/pushing that to it's own tag until you're confident that this rewrite is fully what we want. Or alternatively keep the python build under latest and push the rust build to a `rust` tag until it's ready for prime time. Whatever you think is best I'm good with.

I've developed a test suite to verify program logic and network connectivity across UDP, TCP and ZMQ. The test suite passes and runs as part of the on pr action. The test suite tests program logic (correct number of messages are sent and received, dedupe, adding or not the proxy id, station id changing, and verifying the data is not mangled otherwise) with UDP only. Other network stacks only check deduplication and correct number of messages. We may want to expand the test suite to also validate that multiple inputs / consumers on many types of inputs all work (ie, does it correctly send data both on UDP and TCP, etc) but for now that case is not covered.

There is an issue related to ARM builds and Rust, at least on GitHub's CI. Basically, cargo will not update/build correctly because of a qemu bug that seems to affect a small number of users on Github and we just happen to be in that group. This bug has been around forever the best I can tell, and again as best as I can tell neither Rust or the qemu people are going to fix it because they seem to regard the problem as "the other guys". As a result we need to run a special step in the build process to increase tmpfs disk size. The on_pr action does this already, but the deploy action doesn't. I will need to add this as an optional configuration item for the deploy action in the common github action repo, as well as add that config option to the deploy action in this branch before we can merge in to main. I'll add in that PR number to the first message of this PR when I get that done.

Also, and I'm not sure if it's tmpfs related or not, but the arm builds are **SLOW**. Like really slow. Honestly, even the AMD64 builds are pretty slow....and the build on my M1 Mac is slow....but I guess that's kind of a Rust thing. :/ At least the resulting binary is quick :)

As far as code quality.....well, I think it's pretty good for a start, but this is my first go at writing Rust (literally never have even done a hello world app; the first time I ran anything was working out how to read in command line args in Rust) so naturally I've probably done things the hard way, inefficiently, or not all "Rusty" like. Also, I've never written async code before in any language so that was a huge learning process and I suspect I've done things not as well as I could have done. I'll keep iterating and updating as I go because I'm quite enjoying learning this.

I've sprinkled in some `TODOs` in the code that are not blocking for functionality, but are things that need to be iterated and fixed ideally.

Suspected issues and pending items: 

* There is probably some wonkyness if there are disconnected TCP sockets and sending data as in it might block the output queue indefinately waiting for it. I haven't tested that fully. 
* UDP packet reconstruction for incoming data is non-existant, but I don't think this is a big issue. It needs to do it, but it shouldn't stop us from testing.
* It doesn't write a stats file. I wasn't sure if this was an important item, but if it is let me know and I'll make it do it.
* Logging output needs to be cleaner so people can provide better info back for debugging purposes

Sorry this has taken me so long. Life caught up to me, and I ran in to some Rust learning curve issues that threw me for a loop and I was discouraged for a while. In fact, it keeps deciding to be....interesting.

Lastly, as we discussed ages ago via DM if this rust rewrite isn't what you want, I'm totally okay with that. I'm sure once you look at this code, and the syntax and all that, you may at least feel like changing your mind :) Or if we want extensive testing before this goes live, that's cool too. Whatever you think is best works for me.

File: .github/workflows/on_pr.yaml [hunk 1]
```diff
@@ -40,6 +42,34 @@ jobs:
       - name: Run hadolint against Dockerfiles
         run: docker run --rm -i -v "$PWD":/workdir --workdir /workdir --entrypoint hadolint hadolint/hadolint --ignore DL3013 --ignore DL3008 $(find . -type f -iname "Dockerfile*")
 
+  # test_rust_functionality:
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply .github/workflows/on_pr.yaml:45]: Ideally yes, but for some reason GitHub decided it doesn't like my test suite. It ends getting hung up on one of the TCP tests and never exits. It runs fine on my Mac. I haven't looked hard at it yet

File: .github/workflows/pre-commit-updates.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+name: Update pre-commit hooks
+
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply .github/workflows/pre-commit-updates.yaml:2]: For my sanity I added a bunch of pre-commit hooks (https://pre-commit.com) that run every time I commit. It does things like listing, makes sure the Rust code is actually valid (does it compile, mostly) and things like that. This action checks the `.pre-commit-config.yaml` file and makes sure all of the hooks used in it are up to date.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
+repos:
+  # lint yaml, line and whitespace
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.3.0
+    hooks:
+      - id: check-yaml
+        exclude: ^(rootfs/webapp/|acarshub-typescript/node_modules/)
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply .pre-commit-config.yaml:7]: Yes. Let me clean this file up a bit. I copied it over and didn't think too hard about it :)

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "acars_router"
+version = "1.0.0"
+edition = "2021"
+authors = ["Fred Clausen", "Mike Nye"]
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply Cargo.toml:5]: :)

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,74 @@
+repos:
+  # lint yaml, line and whitespace
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply .pre-commit-config.yaml:2]: This is the configuration file for the pre-commit hooks.

File: Dockerfile [hunk 6]
```diff
@@ -7,33 +17,16 @@ ENV AR_LISTEN_UDP_ACARS=5550 \
...
-        "${KEPT_PACKAGES[@]}" \
-        "${TEMP_PACKAGES[@]}" \
-        && \
-    python3 -m pip install --no-cache-dir --upgrade pip && \
-    python3 -m pip install --no-cache-dir --requirement /opt/acars_router/requirements.txt && \
-    # Clean up
-    apt-get remove -y "${TEMP_PACKAGES[@]}" && \
-    apt-get autoremove -y && \
-    rm -rf /src/* /tmp/* /var/lib/apt/lists/* && \
+    apt-get install -y --no-install-recommends libzmq3-dev && \
```
[the remark this answers is not in the document]
Comment (fredclausen) [reply Dockerfile:28]: Probably yes. I forget why I left it there. As I recall it needed the ZMQ libraries still installed in the system and uninstalling that package removed them.

Looking at https://packages.debian.org/bullseye/libzmq3-dev it seems like [libzmq5](https://packages.debian.org/bullseye/libzmq5) is installed as part of that package so perhaps installing that separately and keeping it will work. I'll play with it.

**Comment by fredclausen** [comment]
Mike, for versioning of the container is there a method to pass in a version to the common workflow? I see stuff in there about it but it's not really clear to me how it all works.

**Comment by fredclausen** [comment]
Some final thoughts.....kind of a TL;DR of my huge (now hidden by GitHub in the many commits I've done) wall of text:

(I'll do my best to be brief. It's not my forte :) )

* Unit testing passes, but I'd honestly like to see it audited by someone who isn't me, and possibly even scrap the entire thing and it do it again. Ideally, the logic checks (does it correctly handle adding / not adding proxy ID, replacing or not the station id, etc) should be done by Rust itself using Rust's built in ability to have in-place unit tests. Other checks related to network connectivity may not even need to be checked if we write Rust tests good enough. I don't know, though

* Ideally I'd like to build this in its own build step and then use the artifact from the build step later on. If that is too much of a pain, then it would be nice to at least have this publish the built version of acars_router for each build target as a release or whatever GitHub calls them so that upstream projects (like ACARS Hub and I presume Plane Watch's docker container?) could grab a binary and go.

* There are some probably overly hack-y bits in here, specifically related to how the message reassembly in UDP/TCP works. To make it work as it should I'll have to re-do some of the flow of the program flow. I'm thinking about how I'd like to accomplish that (while writing this I had a brain storm....) but I don't want to mess with it now.

Basically, I think this is ready for testing. I've had this running in my ACARS stack now for a few days and I've had zero issues. I think it's best that I don't add in new features or change major chunks of code. Those kinds of things, absent a necessity to fix bugs, should be ideally left for a future revision of acars_router.

Thanks for taking a look at this Mike.

File: .github/workflows/pre-commit-updates.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikenye) [reply .github/workflows/pre-commit-updates.yaml:2]: That’s cool!

**Comment by fredclausen** [comment]
@mikenye I wanted to get your thoughts about message reassembly. Right now, with TCP (I'm not sure this is necessary but I added it there) and UDP the program is capable of reassembling messages sent over multiple packets.

Two major limitations:

* It does not factor in message order. It's literally only checking message reassembly in the order the packets are received
* It does not factor in a peer sending multiple split messages at the same time

I've got a partial implementation that takes care of that. I am going to hate writing the tests for this, but, it's def technically possible to deal with those potential cases.

However, I'm not sure we want to. In fact, I'm even questioning the notion of packet reassembly at all. Basically, how can this program know, with just the packets from a single peer, what packet belongs to what message? I'm using the idea that if it deserializes in to valid JSON it must be good, but that seems like a really crap way of doing it.

Take the following example

Message 1
```json
{
   "field1": 1,
   "field2": 2
}
```

Message 2
```json
{
   "field1": 3,
   "field2": 4
}
```

Say the peer split the packet in the middle of field2, like so

Message 1, Packet 1
```
{
   "field1": 1,
   "fiel"
```

Message 1, Packet 2
```
d2": 2
}
```

Message 2, Packet 1
```
{
   "field1": 3,
   "fiel
```

Message 2, Packet 2
```
d2": 4
}
```

Based on packet reception order we could end up thinking these were the messages sent:

Message 1
```json
{
   "field1": 1,
   "field2": 4
}
```

Message 2
```json
{
   "field1": 3,
   "field2": 2
}
```

Both are valid JSON, and it's possible that if the message is split in the right place the resulting deserialization looks like valid JSON but:

* Isn't valid because the field names are mangled
* Data that was reconstructed in the second and subsequent packets are totally wrong.

Basically.....the message integrity, ie knowing that what we made from the packets is what the peer sent can ONLY be known if the entire message is received at once. Never otherwise. I hope my thought process is clear enough.

TL;DR: Should we even be doing message reconstruction at all?

**Comment by mikenye** [comment]
> Mike, for versioning of the container is there a method to pass in a version to the common workflow? I see stuff in there about it but it's not really clear to me how it all works.
> 
> ETA: My thinking is it would be nice to version the container based on the version present in `cargo.toml`

Currently the [`get_version_method`](https://github.com/sdr-enthusiasts/common-github-workflows#get_version_method) only supports:

* Getting the version from a file in the container (eg: `get_version_method: file_in_container:file=/IMAGE_VERSION`)
* Git commit hash (eg: `get_version_method: git_commit_hash_short`)

We should add the ability to get a version from a `cargo.toml` file. I'll open an issue and try to implement, although I will add the caveat that I have very little free time at the moment :(.

**Comment by mikenye** [comment]
> * Unit testing passes, but I'd honestly like to see it audited by someone who isn't me, and possibly even scrap the entire thing and it do it again. It's exceptionally hacky and I feel like the potential for stupid issues to go unnoticed because of bugs in the python tester are too high. Potentially a larger sample of messages, or messages that should cause the message parser to reject it (no time stamp, for instance), multiple peers for input and output, should be included. Ideally, the logic checks (does it correctly handle adding / not adding proxy ID, replacing or not the station id, etc) should be done by Rust itself using Rust's built in ability to have in-place unit tests. Other checks related to network connectivity may not even need to be checked if we write Rust tests good enough. I don't know, though

@jcdeimos are you able to take a look? You're the most proficient rustacean I know, outside of @fredclausen :)

File: src/acars_router_servers/sender_servers.rs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            let s_type = server_type.clone();
+            tokio::spawn(async move { start_tcp(hostname, s_type, new_state).await });
+        }
+    }
+
+    if should_start_service(config.serve_tcp()) {
+        // Start the TCP servers for {server_type}
+
+        for host in config.serve_tcp() {
+            let hostname = "0.0.0.0:".to_string() + host.as_str();
```
Comment (bluntelk) [inline src/acars_router_servers/sender_servers.rs:72]: Hi there,

Sorry to be a pain - this doesn't look right (either host is badly named - port - or your gonna get some weird hostnames

**Comment by mikenye** [comment]
> * Ideally I'd like to build this in its own build step and then use the artifact from the build step later on. If that is too much of a pain, then it would be nice to at least have this publish the built version of acars_router for each build target as a release or whatever GitHub calls them so that upstream projects (like ACARS Hub and I presume Plane Watch's docker container?) or users who want to run this outside of docker could grab a binary and go without compiling it.

I have no problem with this. Did you want to implement this as part of this PR, or after the fact?

File: src/acars_router_servers/sender_servers.rs (hunk 7 above)
Comment (mikenye) [reply src/acars_router_servers/sender_servers.rs:72]: Thanks for taking a look @bluntelk. @fredclausen FYI this is "Boxie" the discord.

**Comment by mikenye** [comment]
> @mikenye I wanted to get your thoughts about message reassembly. Right now, with TCP (I'm not sure this is necessary but I added it there) and UDP the program is capable of reassembling messages sent over multiple packets.
> 
> Two major limitations:
> 
> * It does not factor in message order. It's literally only checking message reassembly in the order the packets are received
> * It does not factor in a peer sending multiple split messages at the same time
> 
> I've got a partial implementation that takes care of that. I am going to hate writing the tests for this, but, it's def technically possible to deal with those potential cases.
> 
> However, I'm not sure we want to. In fact, I'm even questioning the notion of packet reassembly at all. Basically, how can this program know, with just the packets from a single peer, what packet belongs to what message? I'm using the idea that if it deserializes in to valid JSON it must be good, but that seems like a really crap way of doing it.
> 
> Take the following example
> 
> Message 1
> 
> ```json
> {
>    "field1": 1,
>    "field2": 2
> }
> ```
> 
> Message 2
> 
> ```json
> {
>    "field1": 3,
>    "field2": 4
> }
> ```
> 
> Say the peer split the packet in the middle of field2, like so
> 
> Message 1, Packet 1
> 
> ```
> {
>    "field1": 1,
>    "fiel"
> ```
> 
> Message 1, Packet 2
> 
> ```
> d2": 2
> }
> ```
> 
> Message 2, Packet 1
> 
> ```
> {
>    "field1": 3,
>    "fiel
> ```
> 
> Message 2, Packet 2
> 
> ```
> d2": 4
> }
> ```
> 
> Based on packet reception order we could end up thinking these were the messages sent:
> 
> Message 1
> 
> ```json
> {
>    "field1": 1,
>    "field2": 4
> }
> ```
> 
> Message 2
> 
> ```json
> {
>    "field1": 3,
>    "field2": 2
> }
> ```
> 
> Both are valid JSON, and it's possible that if the message is split in the right place the resulting deserialization looks like valid JSON but:
> 
> * Isn't valid because the field names are mangled
> * Data that was reconstructed in the second and subsequent packets are totally wrong.
> 
> Basically.....the message integrity, ie knowing that what we made from the packets is what the peer sent can ONLY be known if the entire message is received at once. Never otherwise. I hope my thought process is clear enough.
> 
> TL;DR: Should we even be doing message reconstruction at all?

I found that with the (admittedly pretty hacky) message reassembly in the Python version, I would end up seeing messages that would otherwise be rejected by ACARSHub. So it seems that valid messages do indeed get broken up. This may be an issue with the acarsdec/vdlm2dec demoduators, as I doubt they're sent over the airwaves like this. Ideally this should be fixed at the source...

**Comment by fredclausen** [comment]
> So it seems that valid messages do indeed get broken up. This may be an issue with the acarsdec/vdlm2dec demoduators, as I doubt they're sent over the airwaves like this. Ideally this should be fixed at the source...

It's probably regional bias but I haven't seen the reassembler that I do have in there doing anything since I put it in, but I am also running dumpvdl2 with zmq -> acars_router so maybe that is the reason. I do get the impression that other parts of the world do tend to see larger messages.

In any case, it does basic reassembly of sequential packets from any of the peers feeding it data. I did tightened up the time interval it will consider a packet it couldn't initially deserialize from 10 seconds down to the the same 1 second that the default skew window is so we don't run in to any weird issues where it could reassemble and deserialize a message but then reject it because it's too old.

I suppose I'll leave it the way it is now and we can revisit it later if need be.

File: src/acars_router_servers/sender_servers.rs (hunk 7 above)
Comment (fredclausen) [reply src/acars_router_servers/sender_servers.rs:72]: I need my morning coffee but I think this is just a poorly named variable. The "host" in this instance is storing only port values, and it is checked during initial load up before it gets to this code to ensure it (and the other user input) are all valid.

Comment (fredclausen) [reply src/acars_router_servers/sender_servers.rs:72]: Actually that is a lie, at least the second half. There are a few input values I added that I forgot to make sure the input is valid and sanitized.

**Comment by fredclausen** [comment]
> > * Ideally I'd like to build this in its own build step and then use the artifact from the build step later on. If that is too much of a pain, then it would be nice to at least have this publish the built version of acars_router for each build target as a release or whatever GitHub calls them so that upstream projects (like ACARS Hub and I presume Plane Watch's docker container?) or users who want to run this outside of docker could grab a binary and go without compiling it.
> 
> I have no problem with this. Should we to implement this as part of this PR, or after the fact?

Based on my previous experience with something similar to this (ACARS Hub does a build step to share the web code artifact with later steps) and GitHub CI in general it'll probably take me 100 tries to get it right. Might as well do it at a later date.

File: src/acars_router_servers/sender_servers.rs (hunk 7 above)
Comment (jcdeimos) [reply src/acars_router_servers/sender_servers.rs:72]: If you're joining two str values like that, probably want something more like `let hostname: String = format!("0.0.0.0:{}", host.as_str());`

The `format!()` macro is fantastic for doing things like formatting of values.

**Comment by mikenye** [comment]
> running dumpvdl2 with zmq -> acars_router

Come to think of it, I think the only messages I saw that needed multi-part reconstruction were over TCP/UDP. I don't think I ever saw one via ZMQ...

**Comment by mikenye** [comment]
> Based on my previous experience with something similar to this (ACARS Hub does a build step to share the web code artifact with later steps) and GitHub CI in general it'll probably take me 100 tries to get it right. Might as well do it at a later date.

FYI, I'm working on this right now in my `rust_dev_workflow_dev` branch, and yes it's a PITA. :)

**Comment by fredclausen** [comment]
> > running dumpvdl2 with zmq -> acars_router
> 
> Come to think of it, I think the only messages I saw that needed multi-part reconstruction were over TCP/UDP. I don't think I ever saw one via ZMQ...

I'm not sure with ZMQ it is possible to fragment the message. Just like with TCP I'm given to understand it isn't possible to receive the packets out of order.

It's probably worth going in to the various decoders and fixing up their output side to

* Ensure we're always putting a new line between messages (shoot....I need to fix something related to that. I'm only splitting incoming messages on TCP/UDP on new lines, NOT `}{`.....damn it)
* Ideally each packet on TCP/UDP would have some kind of unique ID for the message, as well as a sequence number so that consumers could reassemble in the right order. However, this would break all of the various existing consumers of data.
* Create a uniform buffer output size. I patched acarsdec ages ago to have a 3000 (?? I forget the exact size) byte buffer, and I THINK vdlm2dec was 8000 bytes. I have no idea what dumpvdl2 does.

Or, we have our own SDR-E decoders.

**Comment by mikenye** [comment]
> I noticed. I get super annoyed it isn't possible to prototype and test without a million pushes to GitHub to see what it'll do. Oh well.

I found something cool:

```yaml
    # Allows troubleshooting via SSH
      - name: Setup upterm session
        uses: lhotari/action-upterm@v1
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          limit-access-to-users: mikenye
```

You need to have your SSH public key in your account. If you add this step to your job, it will start an SSH server and stop there. It gives you the SSH command to join into the session in the log. So handy!

> Out of curiosity have you built/run this yet?

Not yet, aiming to do it tonight. Concentrating on getting the workflows sorted out.

File: src/acars_router_servers/sender_servers.rs (hunk 7 above)
Comment (fredclausen) [reply src/acars_router_servers/sender_servers.rs]: > If you're joining two str values like that, probably want something more like `let hostname: String = format!("0.0.0.0:{}", host.as_str());`
> 
> The `format!()` macro is fantastic for doing things like formatting of values.

I'm using the format macro now, thank you for that. This is my first go at writing Rust and I kind of just jumped in head first so obvious things like that I'm not surprised I missed.

**Comment by fredclausen** [comment]
> I found something cool:

That is seriously cool and useful. Noting that for future use :)

File: src/sanity_checker.rs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        is_input_sane = false;
+    }
+
+    match is_input_sane {
+        true => Ok(()),
+        false => Err("Config option sanity check failed".to_string()),
+    }
+}
+
+fn check_ports_are_valid(option_ports: &Option<Vec<String>>, name: &str) -> bool {
```
Comment (jcdeimos) [inline src/sanity_checker.rs:140]: Now that I'm a little more familiar with the code base, I'm wondering if we want to change any of the option switches that _only_ accept ports from `Option<Vec<String>>` to `Option<Vec<u16>>` instead?

This would mean that your max port number cannot be bigger than 65535, as it's the MAX size for a u16, so you only have to check that it's not zero. It'll cut down a bunch of the code needed to validate your config here.

Comment (fredclausen) [reply src/sanity_checker.rs:140]: That makes a ton of sense
